exhort/api/src/server.rs

```diff
@@ -114,6 +118,80 @@ impl From<collectorist_client::Error> for Error {
     }
 }
 
+#[utoipa::path(
+post,
+request_body = AnalyzeRequest,
+    responses(
+        (status = 200, body = RecommendationResponse, description = "Recommended pURLs"),
+    ),
+)]
+#[post("recommend")]
+async fn recommend(
+    state: web::Data<AppState>,
+    request: web::Json<AnalyzeRequest>,
+) -> actix_web::Result<impl Responder> {
+    let mut recommendations = HashMap::new();
+
+    for purl_str in &request.purls {
+        if let Ok(purl) = PackageUrl::from_str(purl_str) {
+            if let Ok(similar_packages) = state
+                .guac_client
+                .intrinsic()
+                .packages(&PkgSpec {
+                    id: None,
+                    r#type: Some(purl.ty().to_string()),
+                    //namespace: None,
+                    //name: None,
+                    namespace: purl.namespace().map(|inner| inner.to_string()),
+                    name: Some(purl.name().to_string()),
+                    version: None,
+                    qualifiers: None,
+                    match_only_empty_qualifiers: Some(false),
+                    subpath: None,
+                })
+                .await
+            {
+                let pattern = Regex::new("\\.redhat-[0-9]+$").unwrap();
+                let mut similar_purls = Vec::new();
+
+                for pkg in similar_packages {
+                    if let Ok(purls) = pkg.try_as_purls() {
+                        for similar_purl in purls {
+                            if let Some(version) = similar_purl.version() {
+                                if pattern.find(version).is_some() {
+                                    if let Some(input_version) = &purl.version() {
+                                        let input_ver = lenient_semver::parse(input_version);
+                                        let similar_ver = lenient_semver::parse(version);
+
+                                        if let (Ok(input_ver), Ok(mut similar_ver)) = (input_ver, similar_ver) {
+                                            // remove the RHT stupid renaming because semver thinks it means pre-release
+                                            // and that breaks stupid comparisions.
+                                            similar_ver.pre = Prerelease::EMPTY;
+                                            if similar_ver >= input_ver {
+                                                similar_purls.push(similar_purl.to_string());
+                                            }
+                                        }
+                                    } else {
+                                        similar_purls.push(similar_purl.to_string());
+                                    }
+                                }
+                            }
+                        }
+                    }
```

Comment:
Am I the only one who's bothered by nested `if let`'s?

Reply:
Probably not. Am I missing an idiom to unnest them?

I have a series of things that can go wrong, and I don't want to bail when the first one does, trying to weave a happy-path through as many of the inputs as I can.

Reply:
Thanks for not commenting on my editorializing comments though :)

Reply:
They were the highlight of the PR!! :)